database/json_file_storage.go

```diff
@@ -0,0 +1,186 @@
+// See the file LICENSE for licensing terms.
+
+package database
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+var _ RelayerDatabase = &JSONFileStorage{}
+
+type chainState map[string]string
+
+// JSONFileStorage implements RelayerDatabase
+type JSONFileStorage struct {
+	// the directory where the json files are stored
+	dir string
+
+	// Each network has its own mutex
+	// The chainIDs used to index the JSONFileStorage are created at initialization
+	// and are not modified afterwards, so we don't need to lock the map itself.
+	mutexes map[ids.ID]*sync.RWMutex
+	logger  logging.Logger
+}
+
+// NewJSONFileStorage creates a new JSONFileStorage instance
+func NewJSONFileStorage(logger logging.Logger, dir string, networks []ids.ID) (*JSONFileStorage, error) {
+	storage := &JSONFileStorage{
+		dir:     filepath.Clean(dir),
+		mutexes: make(map[ids.ID]*sync.RWMutex),
+		logger:  logger,
+	}
+
+	for _, network := range networks {
+		storage.mutexes[network] = &sync.RWMutex{}
+	}
+
+	_, err := os.Stat(dir)
+	if err == nil {
+		// Directory already exists.
+		// Return the existing storage.
+		return storage, nil
+	}
+
+	// 0755: The owner can read, write, execute.
+	// Everyone else can read and execute but not modify the file.
+	err = os.MkdirAll(dir, 0755)
+	if err != nil {
+		storage.logger.Error("failed to create directory",
+			zap.String("dir", dir),
+			zap.Error(err))
+		return nil, err
+	}
+
+	return storage, nil
+}
+
+// Get the latest chain state from the JSON database, and retrieve the value from the key
+func (s *JSONFileStorage) Get(chainID ids.ID, key []byte) ([]byte, error) {
+	mutex, ok := s.mutexes[chainID]
+	if !ok {
+		return nil, errors.Wrap(
+			ErrDatabaseMisconfiguration,
+			fmt.Sprintf("database not configured for chain %s", chainID.String()),
+		)
+	}
+
+	mutex.RLock()
+	defer mutex.RUnlock()
+	currentState := make(chainState)
+	fileExists, err := s.read(chainID, &currentState)
+	if err != nil {
+		s.logger.Error(
+			"failed to read file",
+			zap.String("chainID", chainID.String()),
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	if !fileExists {
+		return nil, ErrChainNotFound
+	}
+
+	var val string
+	if val, ok = currentState[string(key)]; !ok {
+		return nil, ErrKeyNotFound
+	}
+
+	return []byte(val), nil
+}
+
+// Put the value into the JSON database. Read the current chain state and overwrite the key, if it exists
+func (s *JSONFileStorage) Put(chainID ids.ID, key []byte, value []byte) error {
+	mutex, ok := s.mutexes[chainID]
+	if !ok {
+		return errors.Wrap(
+			ErrDatabaseMisconfiguration,
+			fmt.Sprintf("database not configured for chain %s", chainID.String()),
+		)
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	currentState := make(chainState)
+	_, err := s.read(chainID, &currentState)
```

Comment:
can you add a comment either at top of function or near here about if no file previously existed for this `chainID` we won't error and will make a new one